### PR TITLE
[RFC] Vim 7.4.1758, 7.4.1759, 7.4.1692

### PR DESCRIPTION
**Pull request by Shougo**
vim-patch:7.4.1758

Problem:    Triggering CursorHoldI when in CTRL-X mode causes problems.
Solution:   Do not trigger CursorHoldI in CTRL-X mode. Add "!" flag to
            feedkeys() (test with that didn't work though).

 vim-patch:7.4.1759

Problem:    When using feedkeys() in a timer the inserted characters are not
            used right away.
Solution:   Break the wait loop when characters have been added to typebuf.
            use this for testing CursorHoldI.

vim-patch:7.4.1692
Problem:    feedkeys('i', 'x') gets stuck, waits for a character to be typed.
Solution:   Behave like ":normal". (Yasuhiro Matsumoto)

### CODE REVIEW

File: src/nvim/edit.c [hunk 1]
```diff
@@ -606,7 +606,10 @@ static int insert_execute(VimState *state, int key)
   s->c = key;
 
   // Don't want K_EVENT with cursorhold for the second key, e.g., after CTRL-V.
-  did_cursorhold = true;
+  // but not in CTRL-X mode, a script can't restore the state //
+  if (ctrl_x_mode == 0) {
+    did_cursorhold = true;
```
Comment (justinmk) [inline src/nvim/edit.c]: The other changes in the other files in this commit are fine though. Though they are bizarre, and the Vim patch mentions nothing about the purpose of yet another flag "!".

**Comment by justinmk** [comment]
The `	vim-patch: 7.4.1692, 7.4.1759` commit message needs to have 

    vim-patch:7.4.1692
    vim-patch:7.4.1759

tokens somewhere.

**Comment by Shougo** [comment]
@justinmk Thank you.  I have fixed them.

**Comment by Shougo** [comment]
I have fixed the conflict.

Note: I have removed `Test_cursorhold_insert_ctrl_x()`.
Because, this is NA test.

**Comment by Shougo** [comment]
@justinmk It can be merged?

File: src/nvim/edit.c [hunk 2]
```diff
@@ -605,7 +605,6 @@ static int insert_execute(VimState *state, int key)
   InsertState *s = (InsertState *)state;
   s->c = key;
 
-  // Don't want K_EVENT with cursorhold for the second key, e.g., after CTRL-V.
```
Comment (justinmk) [inline src/nvim/edit.c]: why was this deleted?

File: src/nvim/testdir/test_autocmd.vim [hunk 3]
```diff
@@ -6,3 +6,20 @@ func Test_vim_did_enter()
...
+endfunc
+
+func Test_cursorhold_insert()
+  let g:triggered = 0
+  au CursorHoldI * let g:triggered += 1
+  set updatetime=20
+  call timer_start(100, 'ExitInsertMode')
+  call feedkeys('a', 'x!')
+  call assert_equal(1, g:triggered)
+endfunc
```
Comment (justinmk) [inline src/nvim/testdir/test_autocmd.vim]: why wasn't the `Test_cursorhold_insert_ctrl_x()` from the original patch added?

Comment (justinmk) [inline src/nvim/testdir/test_autocmd.vim]: @bfredl This is the patch in question: https://github.com/vim/vim/commit/40b1b5443c88fab77f1f7c6f9e801f7ffdb7e0a8

Should the change to `WaitForChar()` be incorporated somehow in our `timer_due_cb()`?

**Comment by justinmk** [comment]
In the commit message the `vim-patch:..` token should not have whitespace. It should be like this:

    vim-patch:7.4.1758

The `vim-patch.sh` script does this automatically.

File: src/nvim/testdir/test_autocmd.vim (hunk 3 above)
Comment (bfredl) [inline src/nvim/testdir/test_autocmd.vim]: event processing seems to work well in `Ctrl-X` mode, otherwise it would just block. The question is rather: why is `CursorHoldI` triggered in `Ctrl-X` mode

Comment (Shougo) [inline src/nvim/testdir/test_autocmd.vim]: > This is not applicable for Nvim. We always set `did_cursorhold = true` above, unconditionally.

So, I think it is NA.

**Comment by Shougo** [comment]
Fixed.

File: src/nvim/testdir/test_autocmd.vim (hunk 3 above)
Comment (justinmk) [inline src/nvim/testdir/test_autocmd.vim]: @Shougo Thank, I get it now. I was confused. I think this is where that check should go:

https://github.com/neovim/neovim/commit/7da7ff7c5c79dd36e053f5b5e86046838445bb8e#diff-ce1f54750dd541783db1cf9a24120ba8R479

**Comment by justinmk** [comment]
Merged after adding the changes discussed in https://github.com/neovim/neovim/pull/5640#discussion_r91419009

**Comment by Shougo** [comment]
Thank you!
